File: WebAPI/services/validationServices.js
const EmailValidator = require('deep-email-validator');
const PasswordValidator = require('password-validator');
const validator = require('validator')

async function isEmailValid(email) {
    return String(email)
        .toLowerCase()
        .match(
            /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|.(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/
        );
}

function isUsernameValid(username) {
    let isValid = {
        valid: true,
        message: ''
    }
    if (username.length < 6 || username.length > 15) {
        isValid.valid = false;
        isValid.message = "Username length must >= 6 and <= 15";
    }
    if (!validator.matches(username, "^[a-zA-Z0-9_\.]*$")) {
        isValid.valid = false;
        isValid.message = "Username must only contains a-z, A-Z, 0-9, _, and .";
    }
    return isValid;
}

function isPasswordValid(password) {
    let isValid = {
        valid: true,
        message: ''
    }

    const passwordValidator = new PasswordValidator();
    passwordValidator
        .is().min(8)                                    // Minimum length 8
        .is().max(25)                                  // Maximum length 25
        .has().uppercase()                              // Must have uppercase letters
        .has().lowercase()                              // Must have lowercase letters
        .has().digits()                                // Must have at least 2 digits
        .has().not().spaces()                           // Should not have spaces

    let validateResults = passwordValidator.validate(password, {details: true});
    isValid.valid = validateResults.length === 0;

    if (!isValid.valid) {
        for (let result of validateResults) {
            isValid.message += result.message.replace("The string", "Password") + '. ';
        }
        isValid.message = isValid.message.trim();
    }
    return isValid;

}
function isOAuthSignUpDataValid(username) {
    let isValid = {
        valid: true,
        message: ''
    }
    const validUsername = isUsernameValid(username);
    if (validUsername.valid === false) {
        isValid.valid = false;
        isValid.message = validUsername.message;
    }
    return isValid;
}
async function isSignUpDataValid(email, password, username) {
    let isValid = {
        valid: true,
        message: ''
    }
    const validUsername = isUsernameValid(username);
    if (validUsername.valid === false) {
        isValid.valid = false;
        isValid.message = "Please provide a valid username.";
    } else {
        const validEmail = isEmailValid(email);
        if (validEmail.valid === false) {
            isValid.valid = false;
            isValid.message = "Please provide a valid email address.";
        } else {
            const validPassword = isPasswordValid(password);
            isValid.valid = validPassword.valid;
            isValid.message = validPassword.message;
        }
    }

    return isValid;
}
module.exports = {
    isSignUpDataValid,
    isPasswordValid,
    isUsernameValid,
    isOAuthSignUpDataValid,
    isEmailValid
}

File: WebAPI/services/validationServices.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using System.Text.RegularExpressions;

namespace webapi_csharp.services
{
    public class validationServices
    {
        public static bool isEmailValid(string email) {
            string pattern = @"^(([^<>()[\]\\.,;:\s@\""]+(\.[^<>()[\]\\.,;:\s@\""]+)*)|.("".+""))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$";
            return Regex.IsMatch(email.ToLower(), pattern);
        }

        public static (bool valid, string message) isUsernameValid(string username) {
            bool isValid = true;
            string message = "";

            if (username.length < 6 || username.length > 15) {
                isValid = false;
                message = "Username length must >= 6 and <= 15";
            }
            if (!Regex.IsMatch(username, "^[a-zA-Z0-9_\\.]*$")) {
                isValid = false;
                message = "Username must only contains a-z, A-Z, 0-9, _, and .";
            }

            return (isValid, message);
        } 

        public static (bool valid, string message) isPasswordValid(string password) {
            bool isValid = true;
            string message = "";

            var passwordValidator = new PasswordValidator();
            passwordValidator
                .HasMinimumLength(8);           // Minimum length 8
                .HasMaximumLength(25);          // Maximum length 25
                .HasUpperCase();                // Must have uppercase letters
                .HasLowerCase();                // Must have lowercase letters
                .HasDigit();                    // Must have at least 1 digits
                .HasNoSpace()                   // Should not have spaces

            var valiadteResults = passwordValidator.Validate(password)

            if (valiadteResults.Any()) {
                isValid = false;
                message = string.Join(". ", validateResult.Select(result => result.Replace("The string", "Password"))) + ".";
            }

            return (isValid, message);
        }

        public static (bool valid, string message) isOAuthSignUpDataValid(string username) {
            var validUsername = isUsernameValid(username);
            return (validUsername.valid, validUsername.message);
        }

        public static (bool valid, string message) isSignUpDataValid(string email, string password, string username) {
            var validUsername = isUsernameValid(username);
            if (!validUsername.valid) {
                return (false, "Please provide a valid username.");
            }

            var validEmail = isEmailValid(email);
            if (!validEmail) {
                return (false, "Please provide a valid email address.");
            }

            var validPassword = isPasswordValid(password);
            if (!validPassword.valid) {
                return (validPassword.valid, validPassword.message);
            }
        }
    }

    public class PasswordValidator
    {
        private readonly PasswordRules rules;

        public PasswordValidator()
        {
            rules = new PasswordRules();
        }

        public PasswordValidator HasMinimumLength(int length)
        {
            rules.MinLength = length;
            return this;
        }

        public PasswordValidator HasMaximumLength(int length)
        {
            rules.MaxLength = length;
            return this;
        }

        public PasswordValidator HasUpperCase()
        {
            rules.RequireUpperCase = true;
            return this;
        }

        public PasswordValidator HasLowerCase()
        {
            rules.RequireLowerCase = true;
            return this;
        }

        public PasswordValidator HasDigit()
        {
            rules.RequireDigit = true;
            return this;
        }

        public PasswordValidator HasNoSpace()
        {
            rules.RequireNoSpace = true;
            return this;
        }

        public List<string> Validate(string password)
        {
            var errors = new List<string>();

            if (rules.MinLength > 0 && password.Length < rules.MinLength)
            {
                errors.Add($"The string must have at least {rules.MinLength} characters.");
            }

            if (rules.MaxLength > 0 && password.Length > rules.MaxLength)
            {
                errors.Add($"The string must have at most {rules.MaxLength} characters.");
            }

            if (rules.RequireUpperCase && !password.Any(char.IsUpper))
            {
                errors.Add("The string must contain at least one uppercase letter.");
            }

            if (rules.RequireLowerCase && !password.Any(char.IsLower))
            {
                errors.Add("The string must contain at least one lowercase letter.");
            }

            if (rules.RequireDigit && !password.Any(char.IsDigit))
            {
                errors.Add("The string must contain at least one digit.");
            }

            if (rules.RequireNoSpace && password.Any(char.IsWhiteSpace))
            {
                errors.Add("The string must not contain any spaces.");
            }

            return errors;
        }

        private class PasswordRules
        {
            public int MinLength { get; set; }
            public int MaxLength { get; set; }
            public bool RequireUpperCase { get; set; }
            public bool RequireLowerCase { get; set; }
            public bool RequireDigit { get; set; }
            public bool RequireNoSpace { get; set; }
        }
    }
}